crates/pathfinder/src/storage/schema/revision_0010.rs

```diff
@@ -0,0 +1,452 @@
+use crate::storage::schema::PostMigrationAction;
+
+use anyhow::Context;
+use rusqlite::Transaction;
+
+pub(crate) fn migrate(transaction: &Transaction) -> anyhow::Result<PostMigrationAction> {
+    // We need to check if this db needs fixing at all
+    let update_is_not_required = {
+        let mut stmt = transaction
+            .prepare("SELECT sql FROM sqlite_schema where tbl_name = 'starknet_events'")
+            .context("Preparing statement")?;
+        let mut rows = stmt.query([]).context("Executing query")?;
+        // Unwrap is safe because the schema for this table obviously contains more than
+        // zero SQL statements, as can be seen in revision 7 migration.
+        // The first statement of the schema for this table is the creation of the table
+        // which could be missing the crucial action, which is ON DELETE CASCADE.
+        rows.next()?
+            .unwrap()
+            .get_ref_unwrap("sql")
+            .as_str()?
+            .contains("ON DELETE CASCADE")
+    };
+
+    if update_is_not_required {
+        return Ok(PostMigrationAction::None);
+    }
+
+    // When altering a table in a way that requires recreating it through copying and deletion
+    // it is [recommended](https://www.sqlite.org/lang_altertable.html) to:
+    // 1. create the new table with some temporary name
+    // 2. copy the data from the old table
+    // 3. drop the old table
+    // 4. rename the new table
+    // Instead of the opposite:
+    // 1. rename the old table
+    // 2. create the new table with the final name
+    // 3. copy the data from the old table
+    // 4. drop the old table
+    //
+    // Triggers and indexes are dropped with the old table, so they need to be recreated
+    // The virtual table remains unchanged
+    transaction
+        .execute_batch(
+            r"
+            CREATE TABLE starknet_events_v2 (
+                block_number  INTEGER NOT NULL,
+                idx INTEGER NOT NULL,
+                transaction_hash BLOB NOT NULL,
+                from_address BLOB NOT NULL,
+                -- Keys are represented as base64 encoded strings separated by space
+                keys TEXT,
+                data BLOB,
+                FOREIGN KEY(block_number) REFERENCES starknet_blocks(number)
+                ON DELETE CASCADE
+            );
+
+            INSERT INTO starknet_events_v2 (
+                block_number,
+                idx,
+                transaction_hash,
+                from_address,
+                keys,
+                data)
+
+                SELECT starknet_events.block_number,
+                    starknet_events.idx,
+                    starknet_events.transaction_hash,
+                    starknet_events.from_address,
+                    starknet_events.keys,
+                    starknet_events.data
+
+                FROM starknet_events;
```

Comment:
How does this work with the `starknet_events_keys` FTS5 virtual table?

That one's configured to be an [external content table](https://www.sqlite.org/fts5.html#external_content_tables) that uses the rowid of the events table as a key.. I guess rowids won't match when copying data with the INSERT INTO above.

Reply:
I think I can fix this by _just_ adding rowid to the insert statement, so that the new version of the table is still properly referenced by the virtual one. But this requires another regression test :factory_worker: :hourglass: .